[4.0][RFC] Serialise inline scripts and inline styles

dgrammatiko opened the pull request:
Pull Request for Issue # .

### Summary of Changes
- Drop the separate arrays for inline scripts and styles
- The data will be merged in the relative scrips and styleSheets arrays (can we remove the underscore???)
- Make `JHTML` compatible with these changes

### Testing Instructions
Apply patch and observe the produced head, inline scripts should be between script tags with src, same for the styles
Also the `jQuery.noConflict();` should be inline and not a file... 


### Update to JHtml
for inline script:
```php
JHtml::_('script', '', ['inline' => ‘content for inline script’]);
```
For inline style:
```php
JHtml::_('stylesheet', '', ['inline' => ‘content for inline script’]);
```

### Actual result
Check the structure of the head:
![screen shot 2017-02-24 at 23 56 37](https://cloud.githubusercontent.com/assets/3889375/23322496/eb2dfdc8-faec-11e6-9b4a-75ea79a61df4.png)

### Documentation Changes Required
YUP!

@wilsonge @mbabker @yvesh

dgrammatiko committed:
opps I missed something

yvesh commented on libraries/cms/html/jquery.php:
```diff
@@ -56,7 +56,7 @@ public static function framework($noConflict = true, $debug = null, $migrate = f
 		if ($noConflict)
 		{
-			JHtml::_('script', 'system/jquery-noconflict.min.js', array('version' => 'auto', 'relative' => true));
+			JHtml::_('script', '', [], ['content' => 'jQuery.noConflict();']);
```
Not better to have that in the array params?

array('inline' => true) ?

yvesh replied:
Ah okay you changed it again.. what's the difference to doing addScriptDeclaration?

dgt41 replied:
it's not a file, so we need to append the data in a `<script>` tag.

This is the less intrusive approach (I think for the function)

dgrammatiko committed:
remove jquery-noconflict.js

dgrammatiko committed:
simplify it

to
```php
JHtml::_('script', '', ['inline' => ‘content for inline script’]);
```

yvesh commented on libraries/joomla/document/html.php:
```diff
@@ -164,10 +162,8 @@ public function setHeadData($data)
-		$this->_scripts     = (isset($data['scripts']) && !empty($data['scripts'])) ? $data['scripts'] : $this->_scripts;
-		$this->_script      = (isset($data['script']) && !empty($data['script'])) ? $data['script'] : $this->_script;
+		$this->styleSheets = (isset($data['styleSheets']) && !empty($data['styleSheets'])) ? $data['styleSheets'] : $this->styleSheets;
+		$this->scripts     = (isset($data['scripts']) && !empty($data['scripts'])) ? $data['scripts'] : $this->scripts;
```
cs

yvesh commented on libraries/joomla/document/document.php:
```diff
@@ -133,15 +133,7 @@ class JDocument
-	 * @since  11.1
-	 */
-	public $_script = array();
+	public $scripts = array();
```
A lot of templates use $_script and $_style to unset things like jquery migrate.

dgt41 replied:
Yup, that's the reason I did this as an RFC. There is a B/C break here that needs to be justified

dgt41 replied:
We can still use the vars  
```
public $_script = array();
```
But there gonna be empty, so no clue if that will help. Also there is no $framework anymore due to Mootools removal...

yvesh commented on libraries/joomla/document/renderer/html/head.php:
```diff
@@ -156,7 +157,7 @@ public function fetchHead($document)
 			}
 
-			$buffer .= '<link href="' . $src . '" rel="stylesheet"';
+			$buffer .= $src !== 'inlineStyles_' . $enumarateInlineStyles ? '<link href="' . $src . '" rel="stylesheet"' : '<style>';
```
Can we have that in a jlayout? :-) There is sometimes the need to override something

dgt41 replied:
Sure

mbabker replied:
I'd be careful about putting this stuff into layouts.  At this point you are rendering the HTML for the `<head>` element.  What are you really going to need to be able to override here that should be encouraged through the use of a layout over the existing `onBeforeCompileHead` event?

yvesh replied:
Hmm let me make an example: The way it's rendered now is XHTML (hardcoded) .e.g <link .. />, maybe you  want to use html5 (<link ..>) or looking into the future maybe your output is something totally different (XSLT, XDOC, PDF ..) which requires a different markup of the inclusion of css and script files.

dgt41 replied:
Hmm, I think @mbabker is right here, in the sense that this is the HTML head, so let's stick to the platform (w3c) and HTML5. @yvesh for the other examples you mentioned above I guess we need a pdf, xdoc etc render per se, or am I getting this wrong?

yvesh replied:
Yeah use case is limited, but maybe there will be a new html version coming in 2 years, which uses again a different markup.. i don't know i still like a bit flexibility here.

mbabker replied:
If the standard changes, we address it then.  Any changes would more likely be with the allowed attributes versus changes to the actual markup, and we already allow attributes arrays so it shouldn't be a big deal.

As for the XDOC, PDF, etc. type outputs, use a renderer for those view formats as needed.  This is a renderer specific to the HTML format.  It might be a good base to copy down the road, there might be ways to move some stuff to a trait, but let's keep the scope in this class limited to its current design.

wilsonge commented:
Ok so what's the advantage of this? Like if I was building JDocument from scratch i would probably do something like this. But I don't see the advantage of changing the ordering that inline scripts/styles and files get loaded given how long it's been this way at this point

dgt41 commented:
Two simple examples to point out how the system fails badly with its current architecture:
CSS
com_something, loads:
- bootstrap.css: `.btn { background-color: black;}`
- inline css: `.btn { background-color: black;}`
mod_something loads mod.css with `.btn { background-color: red;}`

Expected colour red, actual black

JS
com_something, loads: 
- core.js Joomla.SubmitForm = function1
- inline css Joomla.SubmitForm = function2
mod_something loads mod.css with Joomla.SubmitForm = function3

Expected function3, actual function2

Both Css and JS need to be appended in the order they had been inserted... :)

dgt41 commented:
I guess this won't go anywhere, closing

dgrammatiko commented:
@wilsonge @yvesh @mbabker can we reconsider this one, if I do it in a B/C way?
Reason: I'm trying to utilise `type="module"` `nomodule` in my apps but I need some inline polyfill for Safari 10.1 which is impossible to do right now without slipping out of the API...

dgrammatiko commented:
Closing, obviously Joomla 4 will be out of sync with the current Javascript and as people keep telling me I shouldn't run after unicorns (in this case js modules)...
